samd: Further extend and improve the samd port.

robert-hh opened the pull request:
This is the second batch, a series of commits, extending and improving somewhat the coverage of MCU features for MicroPython. Additions and improvements:

- Add machine.reset_cause()
- Add machine.lightsleep().
- Enable os.uname()
- Add support for the DHTxx class. Enabled on SAMD51, possible for SAMD21.
- Add support for the Onewire class.  Enabled on SAMD51, possible for SAMD21.
- Add board files for Sparkfun SAMD51 Thing Plus.
- Implement uart.txdone() and uart.flush().
- Add the machine.RTC module. The MCU has a single low power timer only, which can be used either as clock or as counter for deepsleep. I went for RTC, since the lightsleep current is already dominated by the board.
- Change the Pin naming to use GPIO numbers instead of item numbers from pins.csv. That allow using all pins of the MCU, whether specified in pins.csv or not. Pins.csv is still used for assigning names to pins. Pin names like "PA16" or "PC02" can be used as well. 
- Extend machine.ADC with a method adc.read_timed_into(). The sampling   frequency is up to 1 MHz for SAMD51 and up to 350kHz for SAMD21. A callback can be given being called at the end of a sampling run.
- Extend machine.DAC with a method dac.write_timed(). The step frequency   is up to ~500kHz for SAMD51 and ~200kHz for SAMD21. Bursts are supported. A callback can be given being called at the end the output burst.
- Change the SAMD51 us-counter to >60 bit. ticks_us() was incompatible due to it's smaller range. All other ticks besides machine.Timer are now derived from the µs-counter, which removes the slight drift between ticks_ms() and ticks_us().
- Extend the clock frequency range for SAMD51 down to 1 Mhz and to 1 to 48 MHz for SAMD21. That reduces the power consumption if needed, while staying full operational (besides the speed).
- Enable FAT support for SAMD51, allowing to use a SD card through SPI. 
- Add more pins to SEEED WIO Terminal's pins.csv.

- Some minor internal changes.

The LED class is still there, even if the Signal class would be more useful for LEDs. The Signal class is not enabled, but that requires just setting a #define.

I have three more tiny commits around tidying up at soft-reset. The fit on the existing state as well.

robert-hh committed:
samd/onewire: Add the onewire device support class for SAMD51.

robert-hh committed:
samd/modmachine: Add machine.reset_cause().

robert-hh committed:
samd/clock_config: Extend the range of machine.freq().

The value given for machine.freq(f) is extend to the range of
1_000_000 to 200_000_000. Frequencies below 48 MHz will be forced to a
integer fraction of 48 MHZ. At frequencies below 8 MHz USB is switched
off. The power consumption e.g. of a ADAFRUIT_ITSYBITSY_M4_EXPRESS
drops to about 1.5 mA at 1 MHz.

Since the peripheral frequency is dropped as well, timing e.g. of PWM,
UART, I2C and SPI is affected and frequency/baud rate has to set
again after a frequency change below 48 MHz.

robert-hh committed:
samd/clock_config: Support machine.freq() for SAMD21 as well.

The range is 1MHz - 48 MHz. Note, that below 8 MHz there is no
USB support. The frequency will be set to an integer fraction of
48 MHz. And after changing the frequency, the peripherals like
PWM, UART, I2C, SPI have to be reconfigured.
Current consumption e.g. of the Seeed Xiao board at 1 MHz is about
1.5 mA, mostly caused by the on-board LED (green LED with 1k resistor
at 3.3V).

robert-hh committed:
samd/boards: Add board files for Sparkfun SAMD51 Thing Plus.

That device uses an SAMD51J20 MCU with 256k RAM and 1024k flash.

robert-hh committed:
samd/DHT: Add DHTxx sensor support for SAMD51.

robert-hh committed:
samd/storage: Enable FAT support for SAMD51.

Tested with a SD card connected to a SAMD51 board.
The SEEED WIO terminal has a SD-Card reader built-in.

Side change including a separate commit:
- Remove a few obsolete lines from Makefile

robert-hh committed:
samd/modmachine: Implement machine.lightsleep().

Which just sets the CPU clock to 200kHz and switches the peripheral clock
off. Two modes:

machine.lightsleep(duration_ms)

and

machine.lightsleep()

In any mode any configured pin.irq() event will terminate the sleep.

Current consumption in lightsleep for some boards:
1.5 - 2.5 mA when supplied trough an active USB
             (Seeed XIAO w/o power LED, Adafruit ItsyBitsy)
0.8 - 2 mA when supplied through Gnd/+5V (Vusb)
           (Seeed XIAO w/o power LED, Adafruit ItsyBitsy)
< 1 mA for SAMD51 when supplied trough a battery connector.
       (Sparkfun Thing SAMD51 plus)

Related change:

Move the calls to SysTick_Config() into set_cpu_freq().
It is required after each cpu freq change to have ticks_ms run at the
proper rate.

robert-hh committed:
samd/mcu: Get the bootloader magic address from the lib.

Instead of being hard-coded. Then it works for all MCUs.
That fits except for a Sparkfun SAMD51 Thing Plus (known) bug, which uses
192k - 4 as magic address. Therefore, that address is set as well to
avoid a problem when this bug is fixed by Sparkfun.

robert-hh committed:
samd/clock_config: Document the #defines use in clock_init().

Which may be set in the respecitve mpconfigboard.h files.

robert-hh committed:
samd/machine_uart: Implement uart.txdone() and uart.flush().

Using the stream method for uart.flush().
uart.txdone() returns True, if the uart not busy, False otherwise.
uart.flush() waits until all bytes have been transmitted or a timeout
        triggers. The timeout is determined by the buffer size and the
        baud rate.

Fix two inconsistencies when not using txbuf:

- report in ioctl as being writeable if there is room in the tx buffer,
  only if it is configured.
- print the txbuf size if configured.

robert-hh committed:
samd/mcu: Make some settings in mpconfigmcu.h conditional.

- Setting the default for MATH as 1 for both MCU types.

robert-hh committed:
samd/RTC: Add the machine.RTC module.

Methods:
rtc.init(date)
rtc.datetime([new_date])
rtc.calibration(value)

The presence can be controlled by MICROPY_PY_RTC.
If the RTC module is used, the time module uses the RTC as well.

For boards without a 32kHz crystal, using RTC makes no sense, since
it will then use the ULP32K Oscillator, which is not precise at all.
Therefore, it will by default only enabled for boards using a crystal,
but can be enabled in the respective mpconfigboard.h.

robert-hh committed:
samd/Pin: Change the Pin-handling and -numbering.

Pin numbers are now the MCU port numbers in the range

PA0..PA31:  0..31
PB0..PB31: 32..63
PC0..PC31: 64..95
PD0..PD31: 96..127

Pins can be denoted by the GPIO port number, the
name as defined in pins.csv or a string in the
form: Pxnn, like "PA16" or "PD03".

The pins.c and pins.h files are now obsolete. The
pin objects are part of the AF table.

As result of a simplification, the code now supports
using pin names or numbers instead of pin objects
for modules like UART, SPI, PWM, I2C, ADC, pininfo.

robert-hh committed:
samd/clock_config: Extend the SAMD51 us-counter to 60 bit.

- Removing the difference in the time.ticks_us() range between
  SAMD21 and SAMD51.
- Adds the function mp_hal_ticks_us_64() and use it for:
   + SAMD51's mp_hal_ticks_us and mp_hal_delay_us().
     For SAMD21, keep the previous methods, which are faster.
   + mp_hal_ticks_ms() and mp_hal_tick_ms_64(), which saves some bytes
     and removes a potential race condition every 50 days.
- Set the us-counter for SAMD51 to 16 MHz for a faster reading
  of the us value.

Note: With SAMD51, mp_hal_ticks_us_64() has a 60 bit range only,
which is still a long time (~36000 years).

robert-hh committed:
samd/boards: Declare more pins for the SEEED WIO terminal.

Defining all pins from the external 40 Pin connector and
some internal pins like the one for SD and LCD.

robert-hh committed:
samd/print: Print objects in a unique pattern.

All of ADC, DAC, Pin, PWM and SPI looked different.

robert-hh committed:
samd/UART: Use a finaliser to tidy up UART on soft reset.

And use the common sercom_table, saving a few bytes of RAM.

robert-hh committed:
samd/SPI: Enable spi.deinit() and simplify sercom_deinit_all().

sercom_deinit_all() does not need the object pointers.

robert-hh committed:
samd/boards: Rework the pins.csv files.

- Remove the LED_Pxxx definitions from pins.csv, now that
  the LED class is gone.
- Remove the '-' lines
- Add default lines for USB and SWCLK, SWDIO.

mattytrentini commented:
Oh, nice - I'd noticed that, previously, is was necessary to have an `LED_` row in `pins.csv` but now that's not required. Great!

mattytrentini commented:
Have you considered capacitive touch @robert-hh? I have _started_ to look into adding support for it on top of your work but will stop if you're planning on adding it.

I think we can leverage [Adafruit_FreeTouch](https://github.com/adafruit/Adafruit_FreeTouch), as [used in CircuitPython](https://github.com/adafruit/circuitpython/blob/main/ports/atmel-samd/common-hal/touchio/TouchIn.c). Was planning on implementing the same [TouchPad API used in the ESP32](https://docs.micropython.org/en/latest/esp32/quickref.html#capacitive-touch). Not sure yet about how to hook in wake-from-sleep...

robert-hh commented:
> Oh, nice - I'd noticed that, previously, is was necessary to have an LED_

You can always define the GPIO with the LED as Pin. Only then youi have to use the Pin class and not the LED class, which is kind of obsolete anyhow.

robert-hh commented:
> Have you considered capacitive touch @robert-hh? I have started to look into adding support for it on top of your work but will stop if you're planning on adding it.

I considered that. But the TOUCH module is not documented by MicroChip in the way they document the other modules. Just some general description, but no registers and set-up instructions. I looked into the Adafruit driver. They have kind of reverse engineered the support lib of MicroChip and created their own driver, but for SAMD21 only. So I've put it aside. But feel free to add it.
An general Problem with the SAMD21 is lack of flash space. At the moment, 192k is provided for code and 64k for a small flash file system. Of the 192k, ~186 are already used, so there is not much space left. If you drop the math lib, you gain 8k again. Some of the boards have an external flash chip, So that could be used for the default file system instead of the internal flash. That would make the board definition more complicated and split the population between boards with and w/o external flash. At the moment I'm happy that the board defintion files are so simple. So what's on my list for further extensions:

- Support external flash for a file system. Seems to be needed, but makes the scenery complicated.
- I2S support. The chip supports it, but only one of the boards has a built-in SD card as mass storage for sound data. And FAT support is enabled only for the SAMD51, due to lack of flash on the SAMD21.
- MSC support. Easy possible, but not really needed given that mpremote works well and the internal flash file systems are tiny.
- Quadrature encode/Counter. The chips have a good modules for that, but the whole support of it is stuck in MicroPython with the existing PRs for ESP32 and MIMXRT. As long as the approach is not resolved, this has to wait.

So them most urgent task to be done is getting more flash space for the SAMD21. Even if I do not like, that it fragment the board population further and make the board definition files more complicate.
-

robert-hh commented:
I'm about to drop the LED class. It's just a handicapped version of Pin and does not deal with LED polarity. Instead the Signal class can be enabled, which makes proper use of the Pin class and allows to specify the polarity. Any comment @dpgeorge ?

dpgeorge commented:
That sounds very good to me!

robert-hh committed:
samd/Signal: Replace the LED class by the Signal class.

It simplifies and improves the code. The LED_Pxxx lines of the
board.csv lines can still be used, but will be taken as Pin
definitions.

mattytrentini commented:
The SAMD board I was experimenting with was the [Adafruit Neo Trinkey](https://learn.adafruit.com/adafruit-neo-trinkey) - hence why touch was important - there's not much else! 😄 I'll take a look and see how challenging it'll be to implement given the space constraints.

BTW, NeoPixels (and so bitstream), worked fine - after adding `neopixel.py` to the manifest file.

I completely agree with your concerns about flash; 192KiB is getting tight when we're adding many of the popular features...

robert-hh commented:
> Not sure yet about how to hook in wake-from-sleep...

If it is about lightsleep: The actual imlpementation of lightsleep looks for a global bool variable called `EIC_occured`. If that is  set to `true`, lightsleep terminates.
If it's about the module itself, you might have to use an interrupt. You can also hook into the Systick interrupt function in samd_isr.c. But do not add much code there. It's called every ms.

The Adafruit Neo Trinkey is a cute board. If you have board definitions, you may make a PR with them. For boards like these, it would also be fine if the devices could emulate a Keypad or Mouse.

robert-hh commented:
Just a few features sizes that I have collected in the last weeks. 

async :  ~1.5 k
unicode: ~2 k
simple float: ~12k
math lib: ~8k
framebuf: 4k
onewire and DHTxx with dht.py and ds18x20.py_: 3.652 
onewire and DHTxx w/o python scripts: +786 
Uart TX buffer: +308 Bytes (plus txbuf sized RAM)
RTC: ~700 (SAMD51) ~500 (SAMD21)
Signal: 500 bytes.
LED: 400 Bytes
Everything features: +2.3K (SAMD51)
Complex math: +2.5k (SAMD51)

mattytrentini commented:
>The Adafruit Neo Trinkey is a cute board. If you have board definitions, you may make a PR with them

See #9561. I found that per-board manifest files didn't work, ie in that PR `neopixel.py` does _not_ get frozen in the firmware. It _is_ included if `requires('neopixel')` is added to the _port_ manifest file. Is that expected?

>For boards like these, it would also be fine if the devices could emulate a Keypad or Mouse.

Yes, USB HID - triggered from a cap touch pin! - would be a great addition.

>framebuf: 4k

Much as I like `framebuf` it seems like the best bang-for-buck feature to remove from your list for the SAMD21. 

>If it is about lightsleep

Light sleep is handy but deep sleep is more interesting to me; driving the processor into a very low-power mode and waking it with cap touch is a useful combination...

robert-hh commented:
- I am a little bit lost in the manifest jungle. To have a board specific manifest.py, you may have to make it the start of the manifest chain, include boards/manifest.py in it and give that board manifest.py as build option instead of `FROZEN_MANIFEST ?= boards/manifest.py`. 
- Framebuf is by default not included in the SAMD21 version.
- LightSleep on a board like the Trinkey without other components consumes about 0.5 mA. Deepsleep would required using the RTC module, since that is the only counter which runs in hibernation mode and can wake the chip up. But then, the clock cannot be used. Not sure how to proceed. I could implement both and control exclusive access. But even with deepsleep, there may be components on the board that still consume current. For the Seeed XIAO board for instance, this is a power LED, pulling >1mA. So not much gained with deepsleep.

jimmo commented:
> See #9561. I found that per-board manifest files didn't work, ie in that PR `neopixel.py` does _not_ get frozen in the firmware. It _is_ included if `requires('neopixel')` is added to the _port_ manifest file. Is that expected?

> * I am a little bit lost in the manifest jungle. To have a board specific manifest.py, you may have to make it the start of the manifest chain, include boards/manifest.py in it and give that board manifest.py as build option instead of `FROZEN_MANIFEST ?= boards/manifest.py`.

@mattytrentini @robert-hh What Robert said, but rather than making it a build option on the command line you should reference your board's manifest in mpconfigboard.mk -- e.g. https://github.com/micropython/micropython/blob/master/ports/stm32/boards/PYBD_SF2/mpconfigboard.mk#L23

mattytrentini commented:
>...you should reference your board's manifest in mpconfigboard.mk

Ah, thanks @jimmo! That makes sense. Updated and works fine.

>Framebuf is by default not included in the SAMD21 version.

👍🏻 

>LightSleep on a board like the Trinkey... [snip]

That all makes sense, and yes, deep sleep isn't going to work on the Trinkey. I was thinking about it for other samd devices but it's something we can refine if necessary.

robert-hh committed:
samd/PWM: Serialize fast update of settings.

- Any update of freq or duty_cycle requires the previous PWM cycle to
  be finished. Otherwise the new settings are not accepted.

Other changes:
- Report the set duty cycles even when the PWM is not yet started.
- pwm.freq(0) stops the pwm device, instead of raising an expception.
- Clear the duty cycle value cache on soft reset.

dpgeorge commented:
> Enable os.uname()


Please don't enable this.  `sys.implementation` and `sys.version` are the better replacement for `os.uname()`.  Eventually we want to deprecate `os.uname()` so better to just never enable it on `samd`, then no one will miss it.


> Extend machine.ADC with a method adc.read_timed_into()
> Extend machine.DAC with a method dac.write_timed()

These are nice to have, but should be added separately, in a separate PR, so their API can be discussed and deemed appropriate for all ports to try and implement (as was done with `UART.flush()` and `UART.txdone()`).  Having them added in this PR will just hold the other commits up.

robert-hh commented:
sys.version and sys.impementation miss the machine name info. That's why I added it. Any replacement for that? I could differentiate sys.platform to samd21 and samd51 instead of just samd.

adc.read_timed() and dac.write_timed() follow the API of the pyb module methods.
I can take them out (I hope), including the changes for dma_manager and tc_manager.

dhalbert commented:
(This is digressing and could be its own discussion.)

> sys.version and sys.impementation miss the machine name info. That's why I added it. Any replacement for that?

In CircuitPython, we haven't found a substitute for some of the things in `sys.uname`, like the machine name and full version info. We do have libraries that make decisions on things based on the sysname or part of the machine name. It's otherwise hard to find out what machine you're running on. In fact, we added `board.board_id` to help even more.
```py
>>> os.uname()
(sysname='ESP32S2', nodename='ESP32S2', release='8.0.0', version='8.0.0-beta.1-24-g14c9028c1 on 2022-10-10', machine='Adafruit Metro ESP32S2 with ESP32S2')
>>> sys.version
'3.4.0'
>>> sys.implementation
(name='circuitpython', version=(8, 0, 0), mpy=517)
>>> sys.platform
'Espressif'
>>> sys.version_info
(3, 4, 0)
>>> board.board_id
'adafruit_metro_esp32s2'     # This is the `boards/` direectory name
```

dpgeorge commented:
In 402df833fe6da5233c83c58421e81493cda54f67 we added `sys.implementation._machine` (compatible with CPython).  And `sys.version` also now includes the full version string.  Eg:
```
>>> sys.implementation
(name='micropython', version=(1, 19, 1), _machine='Raspberry Pi Pico W with RP2040', _mpy=4358)
>>> sys.version
'3.4.0; MicroPython v1.19.1-468-g0e35c4de9-dirty on 2022-10-05'
>>> sys.platform
'rp2'
>>> sys.version_info
(3, 4, 0)
```
And some ports even have the `platform` module for extra compiler and SDK/IDF info.

robert-hh commented:
OK. So sys,implementation has the machine name.

dhalbert commented:
@dpgeorge Aha, thank you. From https://peps.python.org/pep-0421/#proposal (emphases mine):

> We will add a new attribute to the sys module, called sys.implementation, as an object with attribute-access (as opposed to a mapping). It will contain implementation-specific information.
> 
> The attributes of this object will remain fixed during interpreter execution and through the course of an implementation version. This ensures behaviors don’t change between versions which depend on attributes of sys.implementation.
> 
> The object has each of the attributes described in the [Required Attributes](https://peps.python.org/pep-0421/#required-attributes) section below. **_Those attribute names will never start with an underscore._** The standard library and the language definition will rely only on those required attributes.
> 
> This proposal takes a conservative approach in requiring only a small number of attributes. As more become appropriate, they may be added with discretion, as described in [Adding New Required Attributes](https://peps.python.org/pep-0421/#adding-new-required-attributes).
> 
> While this PEP places no other constraints on sys.implementation, it also recommends that no one rely on capabilities outside those described here. _**The only exception to that recommendation is for attributes starting with an underscore. Implementers may use those as appropriate to store per-implementation data.**_

dpgeorge commented:
See #8572 and #8588 for a much longer discussion on `sys.implementation`.

robert-hh commented:
@dpgeorge So I remove all commits related to adc_timed() and dac_timed(). There are three tiny new commits. The first is another fix for PWM. The other two just change the tidying up at soft_reset(). But I see that the latest merge of @jimmo 's work on Makefile makes trouble.

jimmo commented:
> But I see that the latest merge of @jimmo 's work on Makefile makes trouble.

@robert-hh I have done the rebase of your branch here -- https://github.com/jimmo/micropython/tree/roberthh-samd

robert-hh commented:
@jimmo Thank. But meanwhile I did that as well and verified, that the branch builds.

jimmo commented on ports/samd/mcu/samd51/mpconfigmcu.mk:
```diff
@@ -1,7 +1,16 @@
 
+SRC_C += \
+	fatfs_port.c \
+	extmod/modonewire.c \
```
Please remove extmod/modonewire.c (this is now added by extmod/extmod.mk).

robert-hh replied:
Done

jimmo commented on ports/samd/mcu/samd51/mpconfigmcu.mk:
```diff
@@ -1,7 +1,16 @@
+	extmod/modonewire.c \
+	drivers/dht/dht.c \
+
+SRC_QSTR += \
```
This is no longer required (SRC_QSTR is now automatic).

robert-hh replied:
Done

jimmo commented on ports/samd/mcu/samd51/mpconfigmcu.mk:
```diff
@@ -1,7 +1,16 @@
+SRC_C += \
+	fatfs_port.c \
+	extmod/modonewire.c \
+	drivers/dht/dht.c \
```
And very minor, this should now be in `DRIVERS_SRC_C` instead of `SRC_C`.

robert-hh replied:
I've noticed that already, as it prevented buidling.

robert-hh committed:
samd/SERCOM: Protect SerCom objects from getting released.

Protect SerCom (UART, SPI, I2C) objects from getting released when they
go out of scope without being deinitialized. Otherwise the ISR
of an Sercom may access an invalid data structure.

robert-hh commented:
Just some cleanup. And os.uname() was still there, even if I remember to have it removed.

robert-hh committed:
samd: Add machine.softreset() for consistency.

dpgeorge commented:
Can you please move the `dht_readinto` method to the `machine` module?  Then at least `dht.py` from micropython-lib will work without any changes to that module.  (And this matches how renesas-ra does it.  Eventually we should move all port's `dht_readinto` methods to the machine module, or at least in some other common location.)

robert-hh commented:
Done, by changing the initial commit for DHT.

robert-hh commented:
> Eventually we should move all port's dht_readinto methods to the machine module, or at least in some other common location.)

That requires dht.py in micropython-lib to be changed as well in a backward compatible style. Codewise this is easy, but doing it is hindered at the moment by the fact, that micropython-lib as linked to MicroPython is in a detached head state.

dpgeorge commented:
Merged in edc3f3d0d3c4099dc6bce5371df0e1c3c5a4aecb through fcccfc176b225001b402ce2ced062344990789e0, with minor changes to rename `MICROPY_PY_RTC` to `MICROPY_PY_MACHINE_RTC`, and `MICOPY_PY_DHT` to `MICROPY_PY_MACHINE_DHT_READINTO`, and a few other very minor formatting changes.

Thanks @robert-hh for the work on this!
